Before start (#413)
codewars/docs: content/training/beginners.md

---
title: Before you start
---

There are no strict requirements, or things which you really, really, need to know, before you start solving your first problems on Codewars. However, being familiar with some basic topics can help with a smooth start.

## Architecture of a Codewars kata

From a solving person's perspective, every kata is built of two main parts: tests, and a solution.

Tests are created by a kata author. Their goal is to verify correctness of a user solution and to be able to do so, tests have to be able to interact with it: if a solution is a function, tests have to know how to call it. If solution is a class, tests have to know how to instantiate it. If solution is a variable, tests have to know how to reference it. That's why tests always predetermine what a solution is, and how it should be interacted with. A solution has to be what tests expect it to be, including its general form (a function, a method, a class, a variable, etc.), name, types, amount of parameters, etc.

> Solution is writen by a user who attempts a kata. To work correctly, it has to conform to expectations of tests: it has to be a function if tests expect it to be one, or a class, or a variable. Statement of a kata task should clearly explain what is required. To make things easier for users, initial solution setup usually (but not always) contains a stub which presents an expected form of a solution which conforms to requirements imposed by tests. It is ready to be called, usually should not be changed by users, and just needs to be filled in. Additional hints about how a user solution should look like and how it is invoked can be found in example tests.

Author: ```suggestion
Solution is written by a user who attempts a kata. To work correctly, it has to conform to expectations of tests: it has to be a function if tests expect it to be one, or a class, or a variable. Statement of a kata task should clearly explain what is required. To make things easier for users, initial solution setup usually (but not always) contains a stub which presents an expected form of a solution which conforms to requirements imposed by tests. It is ready to be called, usually should not be changed by users, and just needs to be filled in. Additional hints about how a user solution should look like and how it is invoked can be found in example tests.
```

 - **How functions accept inputs?** - on Codewars, solution functions receive inputs by their parameters. Especially, users are not required to read anything from console or anywhere else.
 - **How functions return results they came up with?** When your function is run, it's expected by tests to return an answer to its caller, so it could be checked and tested. In particular, printing the answer to a console, or anywhere else, will not work as expected.
- **How are functions invoked?** - tests will call your function, possibly multiple times. Every time, they will pass in different input arguments, and will inspect an answer returned by it. There is no need for users to call their function themselves, and there's no need for any kind of entry point (like some `main` function, or equivalent) just to make a solution runnable. Tests already know how to run solutions. Users also need to make sure that their solution can be run more than once, and every time returns a correct answer.

## Solution as a class

> Another popular kind of a solution is a class, or some function associated with a class (like a class method, or a prototype extention, etc.) In majority of cases, it's enough to fill in a solution template in a correct place to create a working solution. But sometimes, some additional knowledge might come in handy:

Author: ```suggestion
Another popular kind of a solution is a class, or some function associated with a class (like a class method, or a prototype extension, etc.) In the majority of cases, it's enough to fill in a solution template in a correct place to create a working solution. But sometimes, some additional knowledge might come in handy:
```

>  - **How are classes constructed and how to create a new one?** - as much as it's helpful to know this in general, this part is usually handled by kata in a solution setup provided to users. In majority of cases a stub of a required classes is already provided and there should be no need to create one from scratch.

Author: ```suggestion
 - **How are classes constructed and how to create a new one?** - as much as it's helpful to know this in general, this part is usually handled by kata in a solution setup provided to users. In the majority of cases a stub of a required class is already provided and there should be no need to create one from scratch.
```
